### ISSUE TITLE
Clarify StackNavigator inside DrawerNavigator gotchas

### ISSUE THREAD

**Issue by brentvatne** [react-navigation/react-navigation.github.io#117]
Moved from https://github.com/react-navigation/react-navigation/issues/3890

<hr />

_Using multiple Navigators inside a DrawerNavigator, where the subnavigators shared a route name, led to unexpected (undocumented) behavior_

| software | version
| -- | --
| react-navigation | 1.5.9
| react-native | expo@26

## Context
I was an early user of @expo/ex-navigation, and have recently been upgrading to react-navigation.  I came across some faults in either the DrawerNavigation documentation or in the general mental framework of react-navigation.

## Current Behavior
When nesting navigators inside a DrawerNavigator, duplicated route names behave unexpectedly (in my mental model).
[Expo Snack repro](https://snack.expo.io/ByWqlS-jG)

```javascript
const WidgetStack = StackNavigator({
  Widget: {
    screen: WidgetScreen,
  },
  // This is very easy to get wrong, it is not clear in the Drawer documentation
  // that this route name should be distinct.
  Settings: {
    screen: SettingsScreen,
  },
});

const SettingsStack = StackNavigator({
  Settings: {
    screen: SettingsScreen,
  },
});

const DrawerApp = DrawerNavigator({
  Widget: {
    screen: WidgetStack,
  },
  Settings: {
    screen: SettingsStack,
  },
}, {
  initialRouteName: 'Widget',
});
```

When tapping on the "Settings" drawer item, I expected the currently selected "tab" in the drawer to be "Settings", I expected no routes to be "pushed" onto my stack navigator.  Instead, the "widget" tab stays selected, and a settings route is pushed onto the WidgetStack.

There were no pieces of documentation to warn me of this behavior, and it was actually very difficult to even tell this is what was happening (it was much easier when I decided to build the minimal repro).

## My current mental model
My mental model for react-navigation was previously that each nested navigator was a "subtree".  With that model, I saw no issue with a duplicated route name.

```
               DrawerApp
	      /         \
	  WidgetStack    SettingsStack
          |     |	       |
	  /      \           Settings
	Widget   Settings
```

I'm not here to debate whether this is the correct model, but a simple piece of documentation would have saved me about an entire day :(

I would be happy to submit a PR to address this, but it appears with the [2.0 push](https://github.com/react-navigation/react-navigation/pull/3618), this might be out of date, and furthermore - I am not sure if my current model is correct enough to even write docs.

**Comment by brentvatne** [comment]
hi there!

the issue here is with route names. unless you have a clear purpose for making multiple routes have the same name, you should always be unique.

so, in this case, what is happening is that when you select the settings tab it dispatches a navigation action to navigate to the route `Settings`. since the deepest active stack navigator can handle the `Settings` route, it pushes the screen to the widgets stack. there are some things we can do to make this behave more sensibly, but avoiding the same routename will resolve the issue here.

**Comment by skabbes** [comment]
Thanks, Brent!

I did already figure it out, the "bug" is not with how it works, but with how it is documented.  I really struggled conceptually with what is a "route", what is a "navigator", and what is a "screen".  And most importantly, how they compose together.

> since the deepest active stack navigator can handle the Settings route
While this makes sense inside a stack navigator, it did not make sense to me when the "active" navigator was the DrawerNavigator.

I didn't see reference to this anywhere in the documentation, not complaining just mentioning.

> unless you have a clear purpose for making multiple routes have the same name, you should always be unique.

This was also due to a complete misunderstanding of "routes".  In my mind, a route and a screen were the same concept.  Coming from the web, a route is the "URL" and the screen is the component rendered.  Relative URLs was the way I expected `navigate` to work, more or less.

`navigate('Settings')` would navigate you to the Settings Screen.  But if you're inside a stack router, it pushes it.  If you are inside a tab, it switches tabs.  Similar to "./settings" or "../settings" potentially on web.

I did **not** expect the router to _reach into_ my routes to try to deep match them, just like on a website - you wouldn't expect ahref="settings" to automatically select /tabs/TabC/settings .  I know web is different from mobile, but when you talk about "navigation" it is surely the most ubiquitous "routing" mental model.

> unless you have a clear purpose for making multiple routes have the same name, you should always be unique.

IMO this really hurts the compose-ability Navigators in react-navigation, as far as I understand it.  It means that routers need to be globally aware of all other routes and be extra careful to not have them overlap.

**Comment by brentvatne** [comment]
hey @skabbes I agree this is a little bit hard to wrap your head around and not well documented. I try to explain how this works in more detail in this talk: https://www.youtube.com/watch?v=wJJZ9Od8MjM - I need to port this to the documentation at some point. related slides: http://url.brentvatne.ca/Vb9cTb

> IMO this really hurts the compose-ability Navigators in react-navigation, as far as I understand it. It means that routers need to be globally aware of all other routes and be extra careful to not have them overlap.

cc @ericvicenti

**Comment by skabbes** [comment]
Awesome, just watched the whole thing.

@brentvatne sounded like you were talking directly to me in minutes 15 to 17 :)

I still don't understand _why_ deepest navigator wins, but doesn't matter - if its documented, I can work with it or work against it!

**Comment by brentvatne** [comment]
@skabbes - it may be useful to imagine what would happen if that wasn't the rule and see how that would play out in various situations. I didn't actually make this design decision but I have found it does make sense in most cases

**Comment by skabbes** [comment]
I mean - I respectfully disagree - given I was making a pretty vanilla app and hit issues due to this behavior.  From your talk, I'm convinced of the merit of "if a route exists, it should be routable from anywhere".  I guess I'm just not sold on the resolution order when multiple routes of the same name exist (which is the only issue I have).

I couldn't really come up with any legitimate use cases for multiple routes of the same name with the current route-resolution behavior (a child-first version of DFS it seems).  Perhaps that should just be an error - and this discussion ends here.  The desire for a global namespace of route names (which seems like a product feature) will always be at odds with maintainability of large apps.  I think a version of BFS that prefers routes in the active route's subtree would "do the right thing" almost always, but that's a major change - and probably not worth it.

**Comment by ericvicenti** [comment]
Thanks for the honest feedback! There are definitely some trade-offs here.

In larger apps I think this route resolution does make sense, because different screens may show up in different contexts, and be linked to from anywhere.

Imagine a settings modal that can come up and have inner navigation to go between the settings screens. Different areas of the app may need to deep link to particular settings screens. So when the user taps a link to one of these specific screens, who is supposed to know which navigator should handle it? The link should be simple, and there may be several contexts where the settings screen can appear. Without this delegation behavior, every single routing behavior would need to be hardcoded within each link!

I'm not really sure what the problem is with using the push action in cases when you want the previous behavior.

Also I'll note that this router delegation behavior is only used as a mechanism of last resort, if a navigator doesn't have a route registered in it. If you don't want to rely on this behavior, you can register all of your routes in the top level navigator and handle things explicitly.

If you want to publish your navigation-aware component to NPM for general consumption, and you want to avoid having conflicting route names with apps that use it, you can use the namespace approach of "my-package/RouteName" to avoid conflicts.

**Comment by ericvicenti** [comment]
In the apps I’ve seen, when you register the same routeName into multiple navigators, you generally want to get the user to the correct place with minimal disruption, so you want the deepest active router to optionally handle the action. This is generally desirable in the UI so the user doesn’t jump to a different area of the app. I can show further examples if you’d like

**Comment by skabbes** [comment]
> you generally want to get the user to the correct place with minimal disruption

Agreed, if not always.  It it at least a sensible default.

> so you want the deepest active router to optionally handle the action

Agree (except maybe in the special case of a custom drawer component). I agreed with this earlier - stated a different way with "[...] a version of BFS that prefers routes in the active route's subtree [...]".

What I disagree with is, when you have to "hop" to a non-active branch of a router (another tab/switch), from there you should prefer the shallowest matching route, not the deepest one.  This corresponds to minimal disruption to the UI.

```javascript
const MyStackNavigator = createStackNavigator({
  Home: HomeStackScreen,
  Profile: ProfileScreen,
});

const MySwitchNavigator = createStackNavigator({
  Home: HomeSwitchScreen,
  Login: SpecialLoginNavigator,
  Main: MyStackNavigator,
});
```

If the Login switch is active, and you `navigate('Home')`, then the SpecialLoginNavigator should handle the action if it can (I agree on this point).  If it cannot... It should do the minimally disruptive thing - and simply switch to HomeSwitchScreen.  It should **not** switch to Main, then push HomeStackScreen.  That would be much more disruptive ¯\\\_(ツ)\_/¯

**Comment by ericvicenti** [comment]
Yeah, I think I agree. Does it currently switch to a deeper "Home" than the topmost-level-one?

Its possible we have a bug here, maybe you could write a failing test test for StackRouter that describes this case?

**Comment by brentvatne** [comment]
@skabbes - in the example you shared:

```
const MyStackNavigator = createStackNavigator({
  Home: HomeStackScreen,
  Profile: ProfileScreen,
});

// there was a typo in your original example where v was createStackNavigator, fixed here
const MySwitchNavigator = createSwitchNavigator({
  Home: HomeSwitchScreen,
  Login: SpecialLoginNavigator,
  Main: MyStackNavigator,
});
```

**Comment by skabbes** [comment]
Thanks, the expo snack I linked should have had all the detail needed too
though.

**Comment by brentvatne** [comment]
@skabbes - my bad, I posted too early and edited a bunch

**Comment by brentvatne** [comment]
https://github.com/react-navigation/react-navigation/blob/ab5481a290a78d75a7bc9f983ced803a0f933ceb/src/routers/__tests__/SwitchRouter-test.js#L186-L216
